### PR TITLE
Add activeUserChart and signupSummary to portal graphql query

### PR DESCRIPTION
**Pull request by carmenlau**
refs #1472 

require #1521

### CODE REVIEW

File: pkg/util/graphqlutil/date.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		return t
+	case *string:
+		if value == nil {
+			return nil
+		}
+		return unserializeDate(*value)
+	case time.Time:
+		// Truncate time
+		t := value.UTC()
+		return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
```
Comment (louischan-oursky) [inline pkg/util/graphqlutil/date.go:42]: Support `*time.Time` as well?

File: pkg/lib/analytic/time_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/authgear/authgear-server/pkg/lib/analytic"
+	"github.com/authgear/authgear-server/pkg/util/periodical"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestGetDateListByRange(t *testing.T) {
```
Comment (louischan-oursky) [inline pkg/lib/analytic/time_test.go:12]: It seems that this function treat rangeTo as inclusive. Would exclusive rangeTo be more conventional? Or we should name the function `GetDateListByRangeInclusive`.

File: pkg/lib/analytic/chart_service.go [hunk 3]
```diff
@@ -14,6 +21,56 @@ func (s *ChartService) GetActiveUserChat(
...
+
+	counts, err := s.AuditStore.GetAnalyticCountsByType(appID, countType, &rangeFrom, &rangeTo)
+	if err != nil {
+		return nil, err
+	}
+
+	countsMap := map[string]int{}
+	dataPoints := []*DataPoint{}
+	for _, c := range counts {
+		countsMap[c.Date.Format("2006-01-02")] = c.Count
```
Comment (louischan-oursky) [inline pkg/lib/analytic/chart_service.go:151]: The constant `"2006-01-02"` is worth defined in `pkg/util/timeutil/layout.go` as

```golang
const LayoutISODate = "2006-01-02"
```

File: pkg/portal/graphql/chart.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+})
+
+var datapoint = graphql.NewObject(graphql.ObjectConfig{
+	Name: "DataPoint",
+	Fields: graphql.Fields{
+		"label": &graphql.Field{
+			Type: graphql.NewNonNull(graphql.String),
+		},
+		"data": &graphql.Field{
+			Type: graphql.NewNonNull(graphql.Int),
```
Comment (louischan-oursky) [inline pkg/portal/graphql/chart.go:29]: Not necessarily an integer. Maybe change the type to Float?

File: pkg/portal/graphql/query.go [hunk 5]
```diff
@@ -109,5 +111,55 @@ var query = graphql.NewObject(graphql.ObjectConfig{
 				}).Value, nil
 			},
 		},
+		"activeUserChart": &graphql.Field{
+			Description: "Active users chart dataset",
+			Type:        activeUserChart,
+			Args: relay.NewConnectionArgs(graphql.FieldConfigArgument{
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:117]: Do not need to use NewConnectionArgs here. We do not actually use the `after`, `before`, `first` and `last` arguments added by NewConnectionArgs.

File: pkg/portal/graphql/query.go [hunk 6]
```diff
@@ -161,5 +161,45 @@ var query = graphql.NewObject(graphql.ObjectConfig{
 				return graphqlutil.NewLazyValue(chart).Value, nil
 			},
 		},
+		"signupSummary": &graphql.Field{
+			Description: "Signup summary for analytic dashboard",
+			Type:        signupSummary,
+			Args: relay.NewConnectionArgs(graphql.FieldConfigArgument{
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:167]: No need to use NewConnectionArgs

**Comment by carmenlau** [comment]
Updated!:)

File: pkg/portal/graphql/query.go [hunk 7]
```diff
@@ -109,5 +111,95 @@ var query = graphql.NewObject(graphql.ObjectConfig{
...
+				},
+				"rangeFrom": &graphql.ArgumentConfig{
+					Type: graphql.NewNonNull(graphqlutil.Date),
+				},
+				"rangeTo": &graphql.ArgumentConfig{
+					Type: graphql.NewNonNull(graphqlutil.Date),
+				},
+			},
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				appID := p.Args["appID"].(string)
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:133]: To stay consistent, the appID we accept here should be encoded app ID. That is base64 encoded `App:raw_app_id`.

File: pkg/portal/graphql/query.go [hunk 8]
```diff
@@ -109,5 +111,95 @@ var query = graphql.NewObject(graphql.ObjectConfig{
 				}).Value, nil
 			},
 		},
+		"activeUserChart": &graphql.Field{
+			Description: "Active users chart dataset",
+			Type:        activeUserChart,
+			Args: graphql.FieldConfigArgument{
+				"appID": &graphql.ArgumentConfig{
+					Type:        graphql.NewNonNull(graphql.String),
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:119]: The type should be ID

File: pkg/portal/graphql/query.go [hunk 9]
```diff
@@ -109,5 +111,95 @@ var query = graphql.NewObject(graphql.ObjectConfig{
...
+					return nil, fmt.Errorf("failed to fetch dataset: %w", err)
+				}
+				return graphqlutil.NewLazyValue(chart).Value, nil
+			},
+		},
+		"signupSummary": &graphql.Field{
+			Description: "Signup summary for analytic dashboard",
+			Type:        signupSummary,
+			Args: graphql.FieldConfigArgument{
+				"appID": &graphql.ArgumentConfig{
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:168]: The type should be ID

File: pkg/portal/graphql/query.go [hunk 10]
```diff
@@ -109,5 +111,95 @@ var query = graphql.NewObject(graphql.ObjectConfig{
...
+				"rangeFrom": &graphql.ArgumentConfig{
+					Type: graphql.NewNonNull(graphqlutil.Date),
+				},
+				"rangeTo": &graphql.ArgumentConfig{
+					Type: graphql.NewNonNull(graphqlutil.Date),
+				},
+			},
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				ctx := GQLContext(p.Context)
+				appID := p.Args["appID"].(string)
```
Comment (louischan-oursky) [inline pkg/portal/graphql/query.go:181]: To stay consistent, the appID we accept here should be encoded app ID. That is base64 encoded `App:raw_app_id`.

**Pull request merged**